### ISSUE TITLE
Accuracy outputs conflicting results in a binary classification setting

### ISSUE THREAD

**Issue by ThScheeve** [PyTorchLightning/metrics#782]
## 🐛 Bug

`Accuracy` (functional implementation; most likely the module as well) gives conflicting results. `StatScores` (functional) does give expected results when I follow the docs. 

As so many input types are supported and (therefore) we have so many parameter options/combinations, it is hard to understand when to use which combination of `average`/`reduce`, `multiclass`, `num_classes`.

### To Reproduce

Following the [docs on the multiclass parameter for torchmetrics](https://torchmetrics.readthedocs.io/en/latest/references/modules.html#using-the-multiclass-parameter), and running my code sample:

#### Code sample

```python
# INPUTS FOLLOWING https://torchmetrics.readthedocs.io/en/latest/references/modules.html#using-the-multiclass-parameter

import torch
from torchmetrics.functional import stat_scores, accuracy

# These inputs are supposed to be binary, but appear as multi-class
preds = torch.tensor([0, 1, 0])
target = torch.tensor([1, 1, 0])

# These inputs are binary (as predictions are probabilities)
preds_prob = torch.tensor([0.2, 0.7, 0.3])
target_prob = torch.tensor([1, 1, 0])

print("These lines ...")
print(stat_scores(preds, target, reduce="micro", num_classes=2))
print(stat_scores(preds, target, reduce="micro", num_classes=1, multiclass=False))
print(stat_scores(preds, target, reduce="macro", num_classes=2))
print(stat_scores(preds, target, reduce="macro", num_classes=1, multiclass=False))
print("... should be the same as these lines ...")
print(stat_scores(preds_prob, target_prob, reduce="micro", num_classes=2, multiclass=True))
print(stat_scores(preds_prob, target_prob, reduce="micro", num_classes=1))
print(stat_scores(preds_prob, target_prob, reduce="macro", num_classes=2, multiclass=True))
print(stat_scores(preds_prob, target_prob, reduce="macro", num_classes=1))
print("-" * 50)

print("Try the same with an actual metric; replace `stat_scores` by `accuracy`")
print("These lines ...")
print(accuracy(preds, target, average="micro", num_classes=2))
print(accuracy(preds, target, average="micro", num_classes=1, multiclass=False))
print(accuracy(preds, target, average="macro", num_classes=2))
print(accuracy(preds, target, average="macro", num_classes=1, multiclass=False))
print("... should be the same as these lines ...")
print(accuracy(preds_prob, target_prob, average="micro", num_classes=2, multiclass=True))
print(accuracy(preds_prob, target_prob, average="micro", num_classes=1))
print(accuracy(preds_prob, target_prob, average="macro", num_classes=2, multiclass=True))
print(accuracy(preds_prob, target_prob, average="macro", num_classes=1))
print("-" * 50)
```

The first block is fine and gives the outputs as expected (output omitted). The second block where we replaced `stat_scores` by `accuracy` gives different results, see arrows:

```
tensor(0.6667)
--> tensor(0.5000)
tensor(0.7500)
tensor(0.5000)

tensor(0.6667)
--> tensor(0.6667)
tensor(0.7500)
tensor(0.5000)
```

### Expected behavior

Would expect `accuracy(preds, target, average="micro", num_classes=1, multiclass=False)` to be equal to `accuracy(preds_prob, target_prob, average="micro", num_classes=1)`

NB:

```
print(accuracy(preds,         target,      average="micro", num_classes=1, multiclass=False))  # A
print(accuracy(preds.float(), target,      average="micro", num_classes=1))  # B
print(accuracy(preds_prob,    target_prob, average="micro", num_classes=1))  # C
```

B == C, but A != B, although according to the [docs](https://torchmetrics.readthedocs.io/en/latest/references/modules.html#using-the-multiclass-parameter):

```
We can set `multiclass=False` to treat the inputs as binary - which is the same as converting the predictions to float beforehand.
```

### Environment

- PyTorch Version (e.g., 1.0): 1.10.1
- OS (e.g., Linux): Win 10 x64
- How you installed PyTorch (`conda`, `pip`, source): `pip`
- Build command you used (if compiling from source): N/A
- Python version: 3.9
- CUDA/cuDNN version: 11.3
- GPU models and configuration: GTX 1080 Ti
- Any other relevant information: torchmetrics 0.7.0

**Comment by Borda** [comment]
@stancld mind have look at this one? ^^ :rabbit:

**Comment by stancld** [comment]
@Borda @SkafteNicki 

I've had a look specifically at the problematic case and found out that for:
```python
accuracy(preds, target, average="micro", num_classes=1, multiclass=False)
```

the derived classification mode is: `DataType.MULTICLASS`, even though it is explicitly set `multiclass=False` in the function call. As the result of this discrepancy, the calculation diverges within the [`_acuracy_compute`]( https://github.com/PyTorchLightning/metrics/blob/51d952d3e969d68d966225d7313c15584bec3ffc/torchmetrics/functional/classification/accuracy.py#L115) method
```python
simple_average = [AverageMethod.MICRO, AverageMethod.SAMPLES]
    if (mode == DataType.BINARY and average in simple_average) or mode == DataType.MULTILABEL:
        numerator = tp + tn
        denominator = tp + tn + fp + fn
    else:
        numerator = tp
        denominator = tp + fn
```

**Comment by SkafteNicki** [comment]
Issue will be fixed by classification refactor: see this issue #1001 and this PR #1195 for all changes

Small recap: This issue describe the discrepancy between using `stat_scores` and `accuracy` in the binary case. The following example is provided:
```python
# These inputs are supposed to be binary, but appear as multi-class
preds = torch.tensor([0, 1, 0])
target = torch.tensor([1, 1, 0])

# These inputs are binary (as predictions are probabilities)
preds_prob = torch.tensor([0.2, 0.7, 0.3])
target_prob = torch.tensor([1, 1, 0])
```
both inputs are binary and the expected output is `2/3=0.666` in both cases when evaluation accuracy. Current implementation fails and gives `0.75` because result is wrongly calculated as an average over the 0 and 1 class: `(1/2 + 1/1) / 2 = 0.75`. Using the new `binary_accuracy` function
```python
from torchmetrics.functional import binary_accuracy
binary_accuracy(preds, target)
# tensor(0.6666)
binary_accuracy(preds_prob, target_prob)
# tensor(0.6666)
```
which give the correct result. Issue will be closed when #1195 is merged.
